talk/expert/perfectforwarding.tex

```diff
@@ -106,133 +106,179 @@
+    \begin{cppcode*}{}
+      template <typename T>
+      void f(T&& t) { ... }
+
+      f(4);            // rvalue -> T&& is int&&
       double d = 3.14;
-      func(d);            // lvalue -> T&& is double&
+      f(d);            // lvalue -> T&& is double&
       float f() {...}
-      func(f());          // rvalue -> T&& is float&&
+      f(f());          // rvalue -> T&& is float&&
       std::string s = "hello";
-      g(std::move(s)); // rvalue -> T&& is std::string&&
+      f(s);            // lvalue -> T&& is std::string&
+      f(std::move(s)); // rvalue -> T&& is std::string&&
+      f(std::string{"hello"}); // rvalue -> std::string&&
     \end{cppcode*}
-  \end{block}
+  \end{exampleblock}
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[11]{std::remove\_reference}
-  Type trait to remove reference from a type
+  \frametitlecpp[20]{Forwarding references}
+  \begin{alertblock}{Careful!}
+    \begin{cppcode*}{gobble=2}
+      template <typename T> struct S {
+        S(T&& t) { ... }       // rvalue references
+        void f(T&& t) { ... }  // NOT forwarding references
+      };
+    \end{cppcode*}
+  \end{alertblock}
+  \begin{exampleblock}{Correct version}
+    \begin{cppcode*}{gobble=2}
+      template <typename T> struct S {
+        template <typename U, std::enable_if_t<
+          std::is_same_v<std::decay_t<U>, T>, int> = 0>
+        S(U&& t) { ... } // deducing context -> fwd. ref.
+        template <typename U>
+        void f(U&& t)    // deducing context -> fwd. ref.
+          requires std::same_as<std::decay_t<U>, T> { ... }
+      };
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Perfect forwarding}
+  \begin{block}{\texttt{std::remove\_reference}}
+    \begin{itemize}
+    \item Type trait to remove reference from a type
+    \item If \mintinline{cpp}{T} is a reference type, \mintinline{cpp}{remove_reference_t<T>} is the type referred to by \mintinline{cpp}{T}, otherwise it is \mintinline{cpp}{T}.
+    \end{itemize}
+  \end{block}
   \begin{block}{}
     \begin{cppcode*}{}
       template <typename T>
       struct remove_reference      { using type = T; };
-
       template <typename T>
       struct remove_reference<T&>  { using type = T; };
-
       template <typename T>
       struct remove_reference<T&&> { using type = T; };
+
+      template <typename T>
+      using remove_reference_t =
+        typename remove_reference<T>::type; // C++14
     \end{cppcode*}
   \end{block}
-  If {\ttfamily T} is a reference type, \mintinline{cpp}{remove_reference_t<T>::type} is the type referred to by {\ttfamily T},
-  otherwise it is {\ttfamily T}.
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[11]{std::forward}
-  Keeps references and maps non-reference types to rvalue references
+  \frametitlecpp[11]{Perfect forwarding}
+  \begin{block}{\texttt{std::forward}}
+    \begin{itemize}
+      \item Forward lvalue-/rvalueness
```

Comment:
```suggestion
      \item Forward lvalue/rvalueness
```

Reply:
I think the dash is correct here, as it means: `lvalueness/rvalueness`.

Reply:
ok, I trust you on the English here, it looked weird to me